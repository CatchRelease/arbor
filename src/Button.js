import styled from 'react-emotion';
import PropTypes from 'prop-types';
import { transparentize } from 'polished';
import {
  borderRadius,
  brandFont,
  colors,
  fontWeights,
  lineHeights,
  spacings,
  typography
} from './theme';

const Button = styled.button`
  font-family: ${brandFont};
  font-size: ${typography.text.ui.fontSize};
  border: none;
  color: ${colors.white};
  font-weight: ${fontWeights.medium};
  line-height: ${lineHeights.large};
  outline: none;
  transition: box-shadow 0.3s ease, background 0.2s ease;

  &:enabled {
    cursor: pointer;
  }
`;

Button.CTA = styled(Button.withComponent('button'))`
  background: ${colors.blue};
  border-radius: ${borderRadius.large};
  box-shadow: 0 1px 3px ${transparentize(0.7, colors.blue)},
    inset 0 -1px 0 ${transparentize(0.9, colors.black)};
  padding: ${spacings.small} 0;
  text-transform: uppercase;
  width: 100%;

  &:disabled {
    background: ${colors.grey80};
    box-shadow: none;
  }

  &:hover:enabled {
    background: ${colors.blueLight};
    box-shadow: 0 2px 6px ${transparentize(0.8, colors.blueLight)};
  }

  &:focus:enabled {
    background: ${colors.blue};
    box-shadow: 0 0 0 1px ${colors.white}, ${colors.blueDark} 0 0 0 2px;
  }
`;

Button.CTA.displayName = 'Button.CTA';

Button.CTA.propTypes = {
  disabled: PropTypes.bool
};

Button.CTA.defaultProps = {
  disabled: false
};

export default Button;
