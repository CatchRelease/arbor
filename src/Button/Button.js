import React from 'react';
import PropTypes from 'prop-types';

import ButtonContent from './ButtonContent';
import ButtonSpinner from './ButtonSpinner';
import ButtonText from './ButtonText';
import StyledButton from './StyledButton';

const Button = React.forwardRef(
  ({ children, iconStart, iconEnd, spin, variant, type, ...props }, ref) => (
    <StyledButton
      {...{
        ...props,
        iconStart,
        iconEnd,
        ref,
        type,
        variant,
        className: `button--${variant}`
      }}
      text={children}
    >
      <ButtonSpinner {...{ spin, variant, ...props }} />
      <ButtonContent spin={spin}>
        {iconStart}
        {children && (
          <ButtonText {...{ iconStart, iconEnd }}>{children}</ButtonText>
        )}
        {iconEnd}
      </ButtonContent>
    </StyledButton>
  )
);

Button.propTypes = {
  /**
   * Text to render inside the button.
   */
  children: PropTypes.string,

  /**
   * Disabled state of button
   * */
  disabled: PropTypes.bool,

  /**
   * Property to set for a button to take the full width of it's parent
   * container.
   * */
  fullWidth: PropTypes.bool,

  /**
   * Icon to render after the button text.
   */
  iconEnd: PropTypes.element,

  /**
   * Icon to render before the button text.
   */
  iconStart: PropTypes.element,

  /**
   * Indicate if the spinner should display.
   */
  spin: PropTypes.bool,

  /**
   * Button size
   * */
  size: PropTypes.oneOf(['small', 'medium', 'large', 'jumbo']),

  /**
   * Button type.
   */
  type: PropTypes.string,

  /**
   * Button variant.
   * */
  variant: PropTypes.oneOf(['primary', 'secondary', 'minimal'])
};

Button.defaultProps = {
  children: undefined,
  disabled: false,
  fullWidth: false,
  iconEnd: undefined,
  iconStart: undefined,
  spin: false,
  size: 'large',
  type: undefined,
  variant: 'primary'
};

export default Button;
