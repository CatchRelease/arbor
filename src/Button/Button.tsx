import { ButtonHTMLAttributes, FC, forwardRef, ReactNode } from 'react';
import { BackgroundColorProps } from 'styled-system';

import ButtonContent from './ButtonContent';
import ButtonSpinner from './ButtonSpinner';
import { Size } from './sizes';
import StyledButton from './StyledButton';
import Text from '../Text';
import { Variant } from './variants';
import { Intent } from '../theme/colors/intent';
import { PaletteColor } from '../theme/colors/palette';

type Props = ButtonHTMLAttributes<HTMLButtonElement> &
  BackgroundColorProps & {
    fullWidth?: boolean | boolean[];
    iconStart: ReactNode;
    iconEnd: ReactNode;
    intent?: Intent;
    paletteColor?: PaletteColor;
    size?: Size | Size[];
    spin?: boolean;
    variant?: Variant;
  };

const Button: FC<Props> = forwardRef<HTMLButtonElement, Props>(
  (
    {
      bg = null,
      children,
      disabled = false,
      fullWidth = false,
      iconEnd,
      iconStart,
      intent = 'brand',
      paletteColor = '',
      size = 'large',
      spin = false,
      type,
      variant = 'primary',
      ...props
    },
    ref
  ) => (
    <StyledButton
      {...{
        bg,
        className: `button--${variant}`,
        disabled,
        fullWidth,
        iconEnd,
        iconStart,
        intent,
        paletteColor,
        ref,
        size,
        type,
        variant,
        ...props
      }}
    >
      {spin && <ButtonSpinner {...{ disabled, intent, size, variant }} />}
      <ButtonContent {...{ hasText: !!children, iconStart, iconEnd, spin }}>
        {iconStart}
        {children && (
          <Text color="inherit" fontSize="inherit">
            {children}
          </Text>
        )}
        {iconEnd}
      </ButtonContent>
    </StyledButton>
  )
);

export default Button;
