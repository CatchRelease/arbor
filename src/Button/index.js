export { default } from './Button';
export { Variant as ButtonVariant } from './variants';
