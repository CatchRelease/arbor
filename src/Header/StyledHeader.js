import React from 'react';

import Grid from '../Grid';

const StyledHeader = (props) => <Grid {...props} />;

StyledHeader.defaultProps = {
  as: 'header'
};

export default StyledHeader;
