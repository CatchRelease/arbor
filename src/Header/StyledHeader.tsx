import Grid, { GridProps } from '../Grid';

const StyledHeader = (props: GridProps) => <Grid {...props} />;

StyledHeader.defaultProps = {
  as: 'header'
};

export default StyledHeader;
