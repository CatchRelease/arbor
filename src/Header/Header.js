import React from 'react';
import PropTypes from 'prop-types';

import StyledHeader from './StyledHeader';

const getColumns = (children = []) => {
  const forHeading = 'minmax(auto, 100%)';

  if (!children.length) return forHeading;

  return `${forHeading} repeat(${children.length}, auto)`;
};

const Header = ({ children, ...props }) => {
  const [headingText, ...remainingChildren] = React.Children.toArray(children);

  return (
    <StyledHeader
      alignItems="end"
      gridGap="small"
      gridTemplateColumns={getColumns(remainingChildren)}
      {...props}
    >
      {headingText}
      {remainingChildren}
    </StyledHeader>
  );
};

Header.propTypes = {
  /**
   * Content to be displayed within the StyledHeader. gridTemplateColumns will
   * be determined by the number of children passed in to properly distribute
   * content.
   */
  children: PropTypes.oneOfType([
    PropTypes.node,
    PropTypes.arrayOf(PropTypes.node)
  ]).isRequired
};

export default Header;
