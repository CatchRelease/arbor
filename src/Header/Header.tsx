import { Children, FC } from 'react';

import StyledHeader from './StyledHeader';

type Children = ReturnType<typeof Children.toArray>;

const getColumns = (children: Children = []) => {
  const forHeading = 'minmax(auto, 100%)';

  if (!children.length) return forHeading;

  return `${forHeading} repeat(${children.length}, auto)`;
};

const Header: FC = ({ children, ...props }) => {
  const [headingText, ...remainingChildren] = Children.toArray(children);

  return (
    <StyledHeader
      alignItems="end"
      gridGap="small"
      gridTemplateColumns={getColumns(remainingChildren)}
      {...props}
    >
      {headingText}
      {remainingChildren}
    </StyledHeader>
  );
};

export default Header;
