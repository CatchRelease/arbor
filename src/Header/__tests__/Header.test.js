import React from 'react';

import Button from '../../Button';
import Heading from '../../Heading';
import Header from '../Header';
import createWithTheme from '../../../utils/createWithTheme';

describe('<Header />', () => {
  it('renders with heading and no buttons', () => {
    const tree = createWithTheme(
      <Header>
        <Heading.h1>My Heading</Heading.h1>
      </Header>
    );

    expect(tree).toMatchSnapshot();
  });

  it('renders with heading and a button', () => {
    const tree = createWithTheme(
      <Header>
        <Heading.h1>My Heading</Heading.h1>
        <Button variant="primary">Primary</Button>
      </Header>
    );

    expect(tree).toMatchSnapshot();
  });

  it('renders with heaeding and multiple buttons', () => {
    const tree = createWithTheme(
      <Header>
        <Heading.h1>My Heading</Heading.h1>
        <Button variant="secondary">Secondary</Button>
        <Button variant="primary">Primary</Button>
      </Header>
    );

    expect(tree).toMatchSnapshot();
  });
});
