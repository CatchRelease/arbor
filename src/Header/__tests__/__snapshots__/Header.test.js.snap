// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> renders with heading and a button 1`] = `
.emotion-0 {
  font-family: Motiva Sans;
  color: #0F0F10;
  font-size: 1.5rem;
  font-weight: 700;
  line-height: 1.25;
  margin-bottom: calc(1.5rem * 0.5);
}

.emotion-8 {
  font-family: Motiva Sans;
  box-sizing: border-box;
  display: grid;
  -webkit-align-items: end;
  -webkit-box-align: end;
  -ms-flex-align: end;
  align-items: end;
  grid-gap: 0.75rem;
  grid-template-columns: minmax(auto, 100%) repeat(1, auto);
}

.emotion-6 {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  outline: none;
  position: relative;
  text-align: center;
  font-size: 0.875rem;
  height: 2.5rem;
  padding: 0 1.25rem;
  background-color: #081D17;
  color: #C0F9DA;
  font-weight: 500;
  width: auto;
  line-height: 1;
  border: 1px solid transparent;
  border-radius: 3px;
}

.emotion-6:disabled {
  background: #EBECEC;
  border: 1px solid transparent;
  color: rgba(120,121,125,0.7);
  cursor: not-allowed;
}

.emotion-6:enabled {
  cursor: pointer;
  -webkit-transition: all 0.3s ease;
  transition: all 0.3s ease;
}

.emotion-6 i {
  -webkit-transition: all 0.3s ease;
  transition: all 0.3s ease;
  vertical-align: middle;
}

.emotion-6 i {
  width: 2.5rem;
  font-size: 0.875rem;
}

.emotion-6 i:first-child {
  margin-left: -1.25rem;
}

.emotion-6 i:last-child {
  margin-right: -1.25rem;
}



.emotion-6:enabled:active {
  background-color: #061813;
}

.emotion-6:enabled:focus {
  box-shadow: 0 0 0 3px #B6EFD0;
}

.emotion-6:enabled:hover {
  background-color: #2f413c;
}

.emotion-6::-moz-focus-inner {
  border: 0;
}

.emotion-6:focus {
  border: 1px solid transparent;
}

.emotion-6:focus:hover {
  border: 1px solid transparent;
}

.emotion-6 i {
  color: inherit;
}

.emotion-4 {
  font-family: Motiva Sans;
  box-sizing: border-box;
  display: inline-grid;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  grid-template-columns: 1fr;
  visibility: inherit;
}

.emotion-2 {
  font-family: Motiva Sans;
  color: inherit;
  font-size: inherit;
  line-height: 1.25;
}

<header
  className="emotion-8 emotion-9"
  display="grid"
>
  <h1
    className="emotion-0 emotion-1"
    color="text.dark"
    fontSize="size7"
    fontWeight="bold"
  >
    My Heading
  </h1>
  <button
    className="button--primary emotion-6 emotion-7"
    disabled={false}
    fontWeight="medium"
    size="large"
  >
    <div
      className="emotion-4 emotion-5"
      display="inline-grid"
    >
      <p
        className="emotion-2 emotion-3"
        color="inherit"
        fontSize="inherit"
      >
        Primary
      </p>
    </div>
  </button>
</header>
`;

exports[`<Header /> renders with heading and no buttons 1`] = `
.emotion-2 {
  font-family: Motiva Sans;
  box-sizing: border-box;
  display: grid;
  -webkit-align-items: end;
  -webkit-box-align: end;
  -ms-flex-align: end;
  align-items: end;
  grid-gap: 0.75rem;
  grid-template-columns: minmax(auto, 100%);
}

.emotion-0 {
  font-family: Motiva Sans;
  color: #0F0F10;
  font-size: 1.5rem;
  font-weight: 700;
  line-height: 1.25;
  margin-bottom: calc(1.5rem * 0.5);
}

<header
  className="emotion-2 emotion-3"
  display="grid"
>
  <h1
    className="emotion-0 emotion-1"
    color="text.dark"
    fontSize="size7"
    fontWeight="bold"
  >
    My Heading
  </h1>
</header>
`;

exports[`<Header /> renders with heaeding and multiple buttons 1`] = `
.emotion-0 {
  font-family: Motiva Sans;
  color: #0F0F10;
  font-size: 1.5rem;
  font-weight: 700;
  line-height: 1.25;
  margin-bottom: calc(1.5rem * 0.5);
}

.emotion-12 {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  outline: none;
  position: relative;
  text-align: center;
  font-size: 0.875rem;
  height: 2.5rem;
  padding: 0 1.25rem;
  background-color: #081D17;
  color: #C0F9DA;
  font-weight: 500;
  width: auto;
  line-height: 1;
  border: 1px solid transparent;
  border-radius: 3px;
}

.emotion-12:disabled {
  background: #EBECEC;
  border: 1px solid transparent;
  color: rgba(120,121,125,0.7);
  cursor: not-allowed;
}

.emotion-12:enabled {
  cursor: pointer;
  -webkit-transition: all 0.3s ease;
  transition: all 0.3s ease;
}

.emotion-12 i {
  -webkit-transition: all 0.3s ease;
  transition: all 0.3s ease;
  vertical-align: middle;
}

.emotion-12 i {
  width: 2.5rem;
  font-size: 0.875rem;
}

.emotion-12 i:first-child {
  margin-left: -1.25rem;
}

.emotion-12 i:last-child {
  margin-right: -1.25rem;
}



.emotion-12:enabled:active {
  background-color: #061813;
}

.emotion-12:enabled:focus {
  box-shadow: 0 0 0 3px #B6EFD0;
}

.emotion-12:enabled:hover {
  background-color: #2f413c;
}

.emotion-12::-moz-focus-inner {
  border: 0;
}

.emotion-12:focus {
  border: 1px solid transparent;
}

.emotion-12:focus:hover {
  border: 1px solid transparent;
}

.emotion-12 i {
  color: inherit;
}

.emotion-4 {
  font-family: Motiva Sans;
  box-sizing: border-box;
  display: inline-grid;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  grid-template-columns: 1fr;
  visibility: inherit;
}

.emotion-2 {
  font-family: Motiva Sans;
  color: inherit;
  font-size: inherit;
  line-height: 1.25;
}

.emotion-14 {
  font-family: Motiva Sans;
  box-sizing: border-box;
  display: grid;
  -webkit-align-items: end;
  -webkit-box-align: end;
  -ms-flex-align: end;
  align-items: end;
  grid-gap: 0.75rem;
  grid-template-columns: minmax(auto, 100%) repeat(2, auto);
}

.emotion-6 {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  outline: none;
  position: relative;
  text-align: center;
  font-size: 0.875rem;
  height: 2.5rem;
  padding: 0 1.25rem;
  background-color: #FFFFFF;
  color: #081D17;
  font-weight: 500;
  width: auto;
  line-height: 1;
  border: 1px solid transparent;
  border-radius: 3px;
}

.emotion-6:disabled {
  background: #EBECEC;
  border: 1px solid transparent;
  color: rgba(120,121,125,0.7);
  cursor: not-allowed;
}

.emotion-6:enabled {
  cursor: pointer;
  -webkit-transition: all 0.3s ease;
  transition: all 0.3s ease;
}

.emotion-6 i {
  -webkit-transition: all 0.3s ease;
  transition: all 0.3s ease;
  vertical-align: middle;
}

.emotion-6 i {
  width: 2.5rem;
  font-size: 0.875rem;
}

.emotion-6 i:first-child {
  margin-left: -1.25rem;
}

.emotion-6 i:last-child {
  margin-right: -1.25rem;
}

.emotion-6:enabled {
  border: 1px solid #E0E1E2;
}

.emotion-6:enabled:active {
  color: #061813;
  background-color: #F8F9F9;
}

.emotion-6:enabled:focus {
  box-shadow: 0 0 0 3px #B6EFD0;
}

.emotion-6:enabled:hover {
  border: 1px solid #2f413c;
}

.emotion-6:enabled:hover i {
  color: #081D17;
}

.emotion-6::-moz-focus-inner {
  border: 0;
}

.emotion-6:focus {
  border: 1px solid transparent;
}

.emotion-6:focus:hover {
  border: 1px solid transparent;
}

.emotion-6 i {
  color: #989A9C;
}

<header
  className="emotion-14 emotion-15"
  display="grid"
>
  <h1
    className="emotion-0 emotion-1"
    color="text.dark"
    fontSize="size7"
    fontWeight="bold"
  >
    My Heading
  </h1>
  <button
    className="button--secondary emotion-6 emotion-7"
    disabled={false}
    fontWeight="medium"
    size="large"
  >
    <div
      className="emotion-4 emotion-5"
      display="inline-grid"
    >
      <p
        className="emotion-2 emotion-3"
        color="inherit"
        fontSize="inherit"
      >
        Secondary
      </p>
    </div>
  </button>
  <button
    className="button--primary emotion-12 emotion-7"
    disabled={false}
    fontWeight="medium"
    size="large"
  >
    <div
      className="emotion-4 emotion-5"
      display="inline-grid"
    >
      <p
        className="emotion-2 emotion-3"
        color="inherit"
        fontSize="inherit"
      >
        Primary
      </p>
    </div>
  </button>
</header>
`;
